Add library information to website

SchrodingersGat opened the pull request:
This is a fairly large PR which adds information on the official KiCad libraries to the website

* Information on where to download the libraries
* Improve information on how to contribute
* Add link to libraries on front page
* Add 'Discover' library section
* Add 'Download' library section

Additionally, the [KLC](https://github.com/KiCad/kicad-library/wiki/Kicad-Library-Convention) have been reformatted and will now exist on the KiCad website rather than on the GitHub Wiki.

The KLC have been largely rewritten and correspond to new v3.0.0 KLC release.

Questions:

1. Who is nominally in charge of maintaining the website and managing PRs?
1. Are there any compatibility issues with the changes I have made? In particular I have used a lot of custom shortcodes to get the KLC to look good, and have had one report that they don't render on certain versions of Hugo
1. If the KLC are moved to the website we may need to update them periodically. I assume this will be ok.

Please look over the changes I have made and let me know if you see any issues.

For reference, the discussion that lead to this work can be found [here](https://lists.launchpad.net/kicad-developers/msg30713.html)

SchrodingersGat commented:
@poeschlr yes before you ask, this does *not* have the footprint naming schema. I would like to see that as 3.1.0 as there's still a lot of work there and if we don't launch this it will never go up :)

nickoe commented:
I got thise error messages when building this
```
ERROR: 2017/10/18 17:39:51 shortcode.go:287: F5.1:8: got positional parameter 'Surface'. Cannot mix named and positional parameters
ERROR: 2017/10/18 17:39:53 shortcode.go:287: S4.2:19: unrecognized character in shortcode action: U+007D '}'. Note: Parameters with non-alphanumeric args must be quoted```

nickoe commented on config.toml:
```diff
@@ -22,16 +22,27 @@ pluralizelisttitles = false
+
+[[menu.main]]
+    parent = "Download"
+		name   = "KiCad Libraries"
```
Indentation is stange

nickoe commented on config.toml:
```diff
@@ -22,16 +22,27 @@ pluralizelisttitles = false
+		weight = 600
+
+[[menu.main]]
+	  parent = "Download"
```
Indentation is stange

nickoe commented on config.toml:
```diff
@@ -22,16 +22,27 @@ pluralizelisttitles = false
 
 [[menu.main]]
-    name   = "Download"
+	  name   = "Download"
```
Indentation is stange

marekr commented:
@SchrodingersGat what version of hugo did you use to build out of curiosity?

SchrodingersGat commented:
Thanks @nickoe I have found and fixed those errors.

@marekr - `Hugo Static Site Generator v0.16-DEV`

poeschlr commented:
Version `Hugo Static Site Generator v0.16 BuildDate: 2016-06-06T14:31:27+02:00` on fedora 26 is the one that made trouble with the shortcodes.

marekr commented on config.toml:
```diff
@@ -29,20 +29,20 @@ pluralizelisttitles = false
-		weight = 300
+    name   = "KiCad Libraries"
+    url    = "/libraries/"
+    weight = 300
```
I think having software up top (smaller weight) might be preferable as that's probably the primary thing the user is looking for when clicking on the menu.

marekr commented:
>Who is nominally in charge of maintaining the website and managing PRs?

Me and nick....nick has got the server stuff going and beats me to accepting PRs as most of them come in Euro timezones which I am not. But I accept PRs too.

>Are there any compatibility issues with the changes I have made? In particular I have used a lot of custom shortcodes to get the KLC to look good, and have had one report that they don't render on certain versions of Hugo

Not that I see

>If the KLC are moved to the website we may need to update them periodically. I assume this will be ok.

That's fine. The site is auto-built (if Jenkins and the machine isn't down) from the repo. So its just PRs.


Feedback:

/libraries/ (the path) ends up causing the Download menu item to be selected because that's where the only link is.

I would almost want to change the main menu structure to have a "Libraries" master entry so it ends up like

Libraries V
Download
License
Contribute
Library Conventions

Like this:
https://i.imgur.com/PWWRL9k.png

But I'm afraid of bloating the header a little with another main entry. But it does centralize library info better.


@nickoe what do you think?

SchrodingersGat commented:
I'm happy to use a libraries menu if that is the consensus

marekr commented:
Can you rebase on the latest master and switch to hugo 0.30?

SchrodingersGat commented:
Can do. Do you want me to consolidate under a singe menu also?

marekr commented on layouts/partials/github_release_table_json.html:
```diff
@@ -0,0 +1,29 @@
+{{ $repo := . }}
+{{ $prefix := "https://api.github.com/repos/kicad/" }}
+
+{{ $json := getJSON $prefix $repo "/releases" }}
```
Hmm, the problem here would be we would need to periodically regenerate the website as getJSON is static one-time.

SchrodingersGat replied:
Yep JavaScript is a much better option. The version above is no longer displayed and I will remove the code once a better version is implemented.

marekr commented on layouts/section/klc-history.html:
```diff
@@ -0,0 +1,29 @@
+		<div class="col-lg-8 col-lg-offset-2 {{ if .Scratch.Get "sidemenu" }}col-md-10 col-sm-9{{ else }}col-sm-12{{ end }}">
+			{{ partial "breadcrumb.html" . }}
+
+			<h2>KLC Revision History</h2>
```
I'll just say I'm not a fan of version numbers in a git repo thats immediately published on every commit. This means you have to bump the version manually every time you change something in the KLC if you want any sanity. And how far do you go, fix one typo, go and bump the version?

poeschlr replied:
A new version number is only introduced if the meaning of the rules does change or if new rules are added. (So a typo does not qualify as that.)

marekr replied:
This should be noted somewhere as part of the PR. CONTRIBUTING.adoc is one good file to update.

poeschlr replied:
I would say details about how the version number is derived might be better at the history page it self. (Then the users have every part of this information on one page.)
But you are right a note in the contribution page on how the KLC is managed might be a good idea.

nickoe commented:
Sorry for my slow response, but I habe been quite busy the last month and have a big email backlog for Kicad.  Only now did I read the initial email thread.  I still need to check your pr out locally have have a more through look at it. The idea is great.

nickoe commented:
Could you please try to resolve the config.toml conflict?  I am not exactly sure what menus you want. It conflicts three times when rebasing your banch onto master because of the updates @marekr made :) And when I did I could not easily see how to navigate to the pull request description. Maybe it was just part of some body text?

SchrodingersGat commented:
Yes I will fix this all when I get home tonight

SchrodingersGat commented:
I have fixed conflicts, made changes, and updated to v0.30 Hugo.

I have also consolidated all the library pages under a single menu. Now the breadcrumbs work much better even for the KLC and I that it is a much better look overall.

SchrodingersGat commented:
The Travis errors are weird, it seems that asciidoctor is now producing errors for auto-continuation of list items? It is still rendering correctly, though. How do I get rid of these error messages?

marekr commented:
@SchrodingersGat hugo used to suppress the warnings, now it forwards them as errors.

Its basically completely about levels

>section title out of sequence: expected level 1, got level 2

=== (level 2) but it wanted to see == (level 1)

> list item index: expected 2, got 1

You have silly indexes.....in F1.1 all the list items start with "1."......
If you want autonumbering, remove the 1 and leave the dot.

SchrodingersGat commented:
Ok I will fix that. Anything else to address?

marekr commented:
Undo the change to the download menu that made it a dropdown as its still a dropdown after you moved out "Download libraries"

SchrodingersGat commented:
I have fixed those problems but I am working on addressing a few other issues before it is ready to merge. I'll let you know :)

SchrodingersGat commented:
Through some [truly awful code](https://github.com/SchrodingersGat/kicad-website/blob/736ccbde5535a21d3b551b86f4584ac74bce2bc6/layouts/shortcodes/fp_code.html) I have managed to setup a reasonably good looking set of footprint naming conventions (that can be built upon)

Hugo, Y U MAKE LIFE SO HARD

@poeschlr if you have the time, the connectors still need work!

![selection_009](https://user-images.githubusercontent.com/10080325/31945198-2edd0df2-b91a-11e7-8e02-4377ef84386d.png)
![selection_010](https://user-images.githubusercontent.com/10080325/31945200-2f177ffa-b91a-11e7-97f6-36aa8bcc531e.png)
![selection_011](https://user-images.githubusercontent.com/10080325/31945201-2f63a308-b91a-11e7-9519-aab5986d0565.png)

marekr commented:
Rebase conflcits by the way ^^

SchrodingersGat commented:
@marekr ? I'm not seeing any conflicts any more..

marekr commented:
![image](https://user-images.githubusercontent.com/1838013/31946127-fd054090-b89e-11e7-986c-37ddd560de0d.png)

SchrodingersGat commented:
I'm not seeing that. I also just rebased my branch off kicad/master and did not have any conflicts... What's going on?

![selection_012](https://user-images.githubusercontent.com/10080325/31946323-416ad3fc-b91d-11e7-87e6-ea372213d4d0.png)

marekr commented:
If you switch that "Merge pull request" dropdown to "Rebase" you'll see the github warning ;)

Did you make sure to git fetch kicad/master before trying rebase locally?

SchrodingersGat commented:
Ah I see.

Yes I did `git fetch` first. Perhaps it is because I have merged kicad/master into my branch at some stage?

marekr commented:
Yea merging master into a branch and merging the branch into master is a great way to make history terrifying and create conflicts :/

SchrodingersGat commented:
Having trouble getting it to work on my end. I'll have another try tomorrow.

SchrodingersGat commented:
Oooookay, lesson learned. Not doing that again. I have rebased the entire commit chain, which was *not* fun. But there are no conflicts now :)

marekr commented:
Download libraries button you put in the homepage banner should link to /libraries/download/ not /libraries/

SchrodingersGat commented:
> Download libraries button you put in the homepage banner should link to /libraries/download/ not /libraries/

Fixed

> The right hand nav is also broken on the libraries section page (it should select the active page)

Fixed

> On the download page
> 
> KiCad libraries are community contributed and available for download on GitHub.
> You link to the github page for KiCad but theres all sorts of stuff in there that aren't libraries.
> 
> Also for non-programmers, GitHub is a terrible place to send them, linking to more immediate directions on how to get libraries is more important.

I also have a link to [http://downloads.kicad-pcb.org/libraries/](http://downloads.kicad-pcb.org/libraries/). What better can we do? The GitHub situation is awkward.

With the upcoming v5 rearranging the situation will be much much better. I have added a "future of the libraries" section where this will be much better.

> You know, while splitting up the KLC into many files for sections is ok, I'm kind of not a fan of how small sections get their own page and you "jump" into one section and go back and then select another.
> It kinds of makes it more tedious to view everything at once if you are familiarizing yourself with the KLC.
> 
> I would be happier if we could come up with a new page style where theres a fixed table of content on the left hand side and just one long page on the right like the docs. Now it won't disrupting studying the KLC for someone unfamiliar and wanting to get a first read. But this is a bit of a exercise maybe for later

Done. /libraries/klc now displays a single continuous list of the KLC rules. It makes reading better and you could for example print all the KLC rules to PDF or something.

Clicking on the title of each rule also sends you to a page for that rule (exactly the same content). This will be useful if we improve the github scripts to send users to the appropriate KLC page when an error is detected.

marekr commented:
>I also have a link to http://downloads.kicad-pcb.org/libraries/. What better can we do? The GitHub situation is awkward.


Well there was that page at some point in th ePR files that would getJson from github, that way the releases data from each library could be wrapped and displayed nicely.

SchrodingersGat commented:
Yes, this is what I will implement when we rearrange the libraries. However:

1. The only repo that would work for is `kicad-library`
1. The `.pretty` repos are spread out everywhere (100+ of the darn things) so that will not work
1. The library-download table will be **great** when we organize the repos as follows:
   * kicad-symbols
   * kicad-footprints 
   * kicad-packages3d
   * kicad-templates

So for now, unless you can think of some other better wording, I say we leave as is.

marekr commented:
Perhaps prefer the archive downloads than over the github link? i.e. Mention archives first, github second

SchrodingersGat commented:
Good idea @marekr see what you think now

nickoe commented:
I think this looks mergable to me now, but I am not sure if it is supposed to be merged now, @SchrodingersGat  mentioned earlier that you are still working on it.

SchrodingersGat commented:
I have finished the work I was doing, and I think it is ready to go. 

(the item I was working on was nice rendering of footprint naming conventions, which are now much more readable).

Please let me know if you see any outstanding issues otherwise I'm very excited to see it merged - this has been a **lot** of work :)

poeschlr commented:
I think it would be a good idea to either increase the font-size for the rule names (section headers) or to decrease the size of headers found within the rules.
![screenshot from 2017-10-27 09-54-21](https://user-images.githubusercontent.com/18327350/32093605-3332f88c-bafd-11e7-9e24-c0539d98c4d8.png)

In the resistor and fuse naming convention the reference to the chip packages is missing. 
There is also no link from the smd chip capacitor section back to the general chip package naming convention. (not a problem when all rules are rendered but if viewed section by section it might be a good idea to have a link to the rule that holds more information.)

A naming convention for J Lead packages is missing. (It is similar to the Gull wing convention. So you could change the header to include J Lead)
The reference back to the manufacturer specific naming convention for IC packages and for cap/resistor naming is missing.

In the connector naming convention it currently looks like the orientation stuff is only valid for manufacturer specific footprints. (This is because you removed the first section that was meant as a general explanation.)

I am not really sure it is a good idea to split all connectors of specific functions (USB, ...) into their own section. 
Might be a maintenance nightmare later on. And i am not convinced it is easier to understand than the version i had previously. 
With my version i think it was also easier to understand how such names are build in general. Which means it makes it easier for users to come up with a naming for connectors that are not explicitly listed.

SchrodingersGat commented:
I'll have a look through your comments later on tonight, thanks Rene.

poeschlr commented:
For later reference, i add the link to the original footprint naming convention discussion:
https://github.com/KiCad/kicad-library/issues/1447
Connector Lib naming:
https://github.com/KiCad/Connectors.pretty/issues/83

And the original discussion for KLCv3
https://github.com/KiCad/kicad-library/issues/1622

SchrodingersGat commented:
Rene, I have now made the following improvements:

1. Simplified the connector naming scheme
   * Generic options listed first
   * Other options simplified
1. Improved documenation for field naming (see KLC F2.2) 
   * Improved descriptions
   * Added long-form prefixes (in addition to abbreviated fields)

Please note that we do NOT want to have to list every type of naming convention. We need enough for clarity, and to provide a pattern for users to follow. 

That said, I have taken some of your suggestions on board and think that it has resulted in an improved look. 

One key change: I have now removed the inidividual optional fields from many footprint naming schemes and replaced it with a single [`Modifiers`] field - this can represent multiple fields.

This improves readability, and also means that we do not have to keep it updated with every single possible optional field. Instead, if we decide on a new optional field type, that can be added to the table in F2.2

poeschlr commented:
The problem with the font sizes for headers still exists.

---

The modifiers field is a good idea. (In general) 
But it seems like you interpreted it as being optional. (The EP size is not optional for footprints that have an exposed pad. Or i at least thought that was our consensus in the original discussion.)

And i could not find a place where you defined the field prefixes to be used. (These need to be standardized such that we can reliably write footprint filters.) The order of the fields also needs to be standardized for the same reason.

---

Tantal cap size codes:
We should require that not only the letter code is given but also the name of the code. (AVX or Kemet)
And i think we should define one of them as preferred. (In the example you use Kemet_ but you never mention it in the rule definition.)

---

About the remarks by @jkriege (abbreviations)
Currently you define the letter W as body with in the prefix definition table.
But in the dip naming convention we use it for the lead span. (This is the only inconsistency that we should address. See https://github.com/KiCad/kicad-library/issues/1447#issuecomment-340164112)

---

Regarding the function field for standardized interface connectors:
@evanshultz [voiced concern that contributors (and maintainers) might use different spellings for the same interface.](https://github.com/KiCad/kicad-library/issues/1447#issuecomment-335208463) This is why i originally included [my list](https://github.com/KiCad/kicad-library/issues/1447#issuecomment-335434885). (The current libs are a mess and can not be used as a reference. This list is also here for us maintainers such that we know how to name things when we cleanup libs.)

---

And something completely different. Currently the menu entry is called KLC. I think it would be better not to use an acronym here. I would use Library Convention if at all possible.

poeschlr commented:
Another thing i just fund. (A mistake i made)

The (marked) part of line:
(can have more digits for connectors with `> 100 pins`)

should be change to `> 99 pins` or `≥ 100 pins` (The later might pose problems if the font of the users browser does not support the greater than or equal sign)

poeschlr commented:
Found another thing:
you use `[pin configuration]` for manufacturer specific connectors and `[pin layout]` for the generic naming. (I think both of them should use the same term.)

SchrodingersGat commented:
> The problem with the font sizes for headers still exists.

If I make the text smaller then it doesn't render nicely in the individual KLC pages. Unless you can provide a way to render the header text small size on the main page, but full-size in the individual pages, I want to leave as it is.

> The modifiers field is a good idea. (In general)
> But it seems like you interpreted it as being optional. (The EP size is not optional for footprints that have an exposed pad. Or i at least thought that was our consensus in the original discussion.)
> 
> Edit: Further more the Lead length was supposed to be required for QFN/DFN.

The modifier field(s) **are** optional. This space is for the overriding fields defined in F2.2. I do not want to enumerate each and every option for different footprint type:

1. That's a lot of work
1. We might miss some, or add ones where they're unnecessary
1. When we think of a new option, we have to edit each footprint type where it is needed.

If a user is reading these and we give the impression that we have thought of *every* permutation, then the requirement is on us to write them all out which is tedious and error prone.

Instead, I have now added the `Modifiers` field which serves as a placeholder for these optional parameters that are spelled out in F2.2

Keep in mind that while serving as a guide for library contributions, users are not reading these without any feedback. I think that writing them as a "guide" rather than a set of fully completed rules helps the users and the librarians equally.

> About the remarks by @jkriege (abbreviations)
> Currently we define the letter W as body within the prefix definition table.
> But in the dip naming convention we use it for the lead span. (This is the only inconsistency that we should address. See KiCad/kicad-library#1447 (comment))
> (Edit: Jan is unhappy with this. Have a look at his comment in the discussion thread.)

Jan is right, that's an awkward way to name things. 

I think that the current DIP naming is a weakness. I think we need a prefix other than `W` which we have assigned to `width`.

Perhaps `S` for span? Anyway, once again let's try not to get too caught up in all the details - otherwise we will never post a 3.0.0 version. These can be tweaked further as we learn more.

> And something completely different. Currently the menu entry is called KLC. I think it would be better not to use an acronym here. I would use Library Convention if at all possible.

Good idea. I have fixed this.

> Regarding the function field for standardized interface connectors:
> @evanshultz voiced concern that contributors (and maintainers) might use different spellings for the same interface. This is why i originally included my list. (The current libs are a mess and can not be used as a reference. This list is also here for us maintainers such that we know how to name things when we cleanup libs.)

The standard I have outlined for functional connectors covers this:

![selection_013](https://user-images.githubusercontent.com/10080325/32142993-eab7b90c-bcf6-11e7-8cc3-3cadd88c3b94.png)

It does not have an explicit example for each style:

1. Keeping it generic reduced workload on librarians
1. Too much information can overload user
1. Once the existing footprints have a consistent naming scheme, it is easy for users to replicate this
1. Pull requests have to be checked over anyway

I think that the information provided in the image above is enough. Too much information is going to be overwhelming. 

> Found another thing:
> you use [pin configuration] for manufacturer specific connectors and [pin layout] for the generic naming. (I think both of them should use the same term.)

All instances now read `Pin layout`. Thanks

> Another thing i just found. (A mistake i made)
> 
> The (marked) part of line:
> (can have more digits for connectors with > 100 pins)
> 
> should be change to > 99 pins or ≥ 100 pins (The latter might pose problems if the font of the users browser does not support the greater than or equal sign)

Fixed.

SchrodingersGat commented:
Thanks for the feedback Rene. I am reluctant to spell out each and every possible combination for each footprint type, that will be far too much work for us, too difficult for users to follow, and prone to error.

The best idea is to have a somewhat generic scheme which allows variance, and (importantly) have a consistent set of existing footprints for users to model their own names on.

SchrodingersGat commented:
@nickoe  @marekr I would like to see this merged unless you have any further objections. Rene and I will no doubt be hashing out the finer points for a while to come, but I think functionally this is good to go. Updates can be made as they come.

Thanks

poeschlr commented:
I agree with oliver. From my side this can be merged.

nickoe commented:
Thank you very much all of you. You have spent quite a lot of energy on this. Thank you. I deem it ready to be merged. Feel free to submit follow up fixed.

I have two comments on what is to be merged that I did not really want to include in the previous complex discussion. This is nitpicking.

1. Does it make sense to "censor" the pull request screenshot on the how to do a github pull request thing?
2. The image showin in libraries/klc/F5.1/ is quite stretched when dispalyed on the page.
